### PR TITLE
feat: add logit_scale to PoolerConfig for affine score calibration

### PR DESCRIPTION
**Pull request by jefp**
## Purpose

Add `logit_scale` to `PoolerConfig` alongside existing `logit_bias`, enabling affine score calibration ([Platt scaling](https://en.wikipedia.org/wiki/Platt_scaling)) in the pooler: `activation(scale * (logit - bias))`.

This allows reranker and classification models to produce calibrated probability scores via `--pooler-config` without custom model code or client-side postprocessing.

RFC: #39434

## Changes

- `vllm/config/pooler.py` — add `logit_scale: float | None = None` field
- `vllm/model_executor/layers/pooler/seqwise/heads.py` — apply scale in `ClassifierPoolerHead.forward()`
- `vllm/model_executor/layers/pooler/seqwise/poolers.py` — pass scale from config
- `vllm/model_executor/layers/pooler/tokwise/heads.py` — apply scale in `TokenClassifierPoolerHead.forward_chunk()`
- `vllm/model_executor/layers/pooler/tokwise/poolers.py` — pass scale from config

**5 files changed, 17 lines added. Zero behavior change for existing models (default is `None`).**

## How it works

```python
# In ClassifierPoolerHead.forward() / TokenClassifierPoolerHead.forward_chunk():
if self.logit_bias is not None:
    logits -= self.logit_bias      # existing
if self.logit_scale is not None:
    logits *= self.logit_scale     # new
if self.activation is not None:
    logits = self.activation(logits)  # existing
```

## Usage

```bash
# No calibration (current behavior, unchanged)
--pooler-config '{"use_activation": true}'

# Bias only (current behavior, unchanged)
--pooler-config '{"use_activation": true, "logit_bias": 0.3}'

# Full Platt scaling (new)
--pooler-config '{"use_activation": true, "logit_bias": 0.3, "logit_scale": 5.0}'

# Temperature scaling (new, special case)
--pooler-config '{"use_activation": true, "logit_scale": 2.0}'
```

## Use cases

1. **Cross-encoder rerankers** — Map raw logits to calibrated `[0,1]` probabilities without retraining
2. **Temperature scaling** — Calibrate overconfident classifiers (`logit_scale = 1/T`)
3. **Score normalization across model versions** — Normalize different model versions to a consistent score range
4. **Domain adaptation** — Calibrate a model for a new domain by fitting two parameters on a small validation set

## Test plan

- [x] Ruff format + check passes
- [x] mypy 3.10 passes
- [x] Tested with cross-encoder reranker model (score calibration produces correct calibrated scores)

```bash
# Test with any cross-encoder
python -m vllm.entrypoints.openai.api_server \
    --model cross-encoder/ms-marco-MiniLM-L-6-v2 \
    --runner pooling \
    --pooler-config '{"use_activation": true, "logit_bias": 0.0, "logit_scale": 1.0}'
```

**Not a duplicate:** No existing PR adds `logit_scale`. This completes the affine calibration pair started by the existing `logit_bias` field.

AI-assisted work: AI assistance was used for implementation. All code reviewed and tested by human submitter.

Cc: @noooop @DarkLight1337

### CODE REVIEW

**Comment by jefp** [comment]
Reopened — the RFC is #39434, this PR implements it.

File: vllm/config/pooler.py [hunk 1]
```diff
@@ -83,6 +83,13 @@ class PoolerConfig:
     If provided, apply classification logit biases. Defaults to None.
     """
 
+    logit_scale: float | None = None
```
[the remark this answers is not in the document]
Comment (jefp) [reply vllm/config/pooler.py:86]: The existing `logit_bias` field is also not included in `compute_hash` — the comment in the method says "this config will not affect the computation graph." Both `logit_bias` and `logit_scale` are simple scalar ops applied after the model forward pass in the pooler head, not part of the compiled computation graph. Following the existing convention here.

File: vllm/model_executor/layers/pooler/seqwise/heads.py [hunk 2]
```diff
@@ -140,6 +142,8 @@ def forward(
         # logits shape: [batchsize, num_labels]
         if self.logit_bias is not None:
             logits -= self.logit_bias
+        if self.logit_scale is not None:
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler/seqwise/heads.py:145]: @noooop do you remember we subtract rather than add the logit bias?

Comment (noooop) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: yes

Comment (noooop) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: https://github.com/vllm-project/vllm/pull/24031/changes

logit_bias is half of estimated_max:

Comment (noooop) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: [mxbai-rerank](https://github.com/mixedbread-ai/mxbai-rerank) sigmoid_normalize

```
def sigmoid_normalize(x: np.ndarray, estimated_max: float = 9.0) -> np.ndarray:
    """Sigmoid function with a fixed maximum value.

    Args:
        x: Input array
        estimated_max: Estimated maximum value of the input

    Returns:
        Values between 0 and 1
    """
    x = x - estimated_max / 2
    return 1 / (1 + np.exp(-x))
```

https://github.com/mixedbread-ai/mxbai-rerank/blob/21d9e79f181298b8dd436bef20d7ac3d80643c9a/mxbai_rerank/utils.py#L9-L20

Comment (noooop) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: Sorry, we should use "add the logit bias." It would be best deprecate "logit_bias."

https://github.com/vllm-project/vllm/blob/c48b2b83bd160cd684c8a2357f229e90de99298d/vllm/v1/worker/gpu/sample/logit_bias.py#L216-L219

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: So our implementation was wrong the whole time and no one noticed it? 🤔

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: Ok let's just change it to addition in this PR then

Comment (noooop) [reply vllm/model_executor/layers/pooler/seqwise/heads.py:145]: My bad, I submitted the wrong implementation by mistake. Normally we just add the logit bias.

**Comment by noooop** [comment]
please update docs

https://github.com/vllm-project/vllm/blob/main/docs/models/pooling_models/classify.md#logit-bias

**Comment by jefp** [comment]
The pooler `logit_bias` (subtract) and sampling `logit_bias` (add) are two different features that happen to share the same name.

**Sampling logit_bias** (`vllm/v1/worker/gpu/sample/logit_bias.py` line 218):
```python
logits += bias   # OpenAI API: shift per-token generation probabilities
```

**Pooler logit_bias** (`seqwise/heads.py` line 142, `tokwise/heads.py` line 129):
```python
logits -= self.logit_bias   # Score calibration: center scores before sigmoid
```

The subtract is correct for the pooler context. Verified against all models currently using it:

```
Model                      Reference code                         vLLM config          vLLM computes            Match
─────────────────────────  ─────────────────────────────────────  ───────────────────  ───────────────────────  ─────
mxbai-rerank-base-v2       sigmoid(x - estimated_max/2)           logit_bias=4.5       sigmoid(x - 4.5)        ✓
                           estimated_max=9.0 → bias=4.5
mxbai-rerank-large-v2      sigmoid(x - estimated_max/2)           logit_bias=6.0       sigmoid(x - 6.0)        ✓
                           estimated_max=12.0 → bias=6.0
JinaVL                     model default set in config.py          logit_bias=2.65      sigmoid(x - 2.65)       ✓
```

Reference: mxbai-rerank `sigmoid_normalize` ([source](https://github.com/mixedbread-ai/mxbai-rerank/blob/21d9e79f181298b8dd436bef20d7ac3d80643c9a/mxbai_rerank/utils.py#L9-L20)):
```python
def sigmoid_normalize(x, estimated_max=9.0):
    x = x - estimated_max / 2      # subtract bias
    return 1 / (1 + np.exp(-x))    # sigmoid
```

The original PR #24031 set `logit_bias = estimated_max / 2` and subtracts it — matching the reference implementation. Changing to addition would break these models.

The naming collision with sampling `logit_bias` is unfortunate but the pooler context (score calibration) and sampling context (token probability shifting) serve different purposes with different sign conventions.

**Comment by jefp** [comment]
Done. Updated docs at `docs/models/pooling_models/classify.md` — added `logit_scale` documentation, computation order, and clarified that `logit_bias` is subtracted (consistent with `sigmoid_normalize` convention).

File: docs/models/pooling_models/classify.md [hunk 3]
```diff
@@ -267,9 +267,30 @@ You can modify the `problem_type` via problem_type in the Hugging Face config. T
 
 Implement alignment with transformers [ForSequenceClassificationLoss](https://github.com/huggingface/transformers/blob/57bb6db6ee4cfaccc45b8d474dfad5a17811ca60/src/transformers/loss/loss_utils.py#L92).
 
-### Logit bias
+### Logit bias and logit scale
```
Comment (noooop) [inline docs/models/pooling_models/classify.md]: How about changing the title to "Affine Score Calibration"?

File: docs/models/pooling_models/classify.md [hunk 4]
```diff
@@ -267,9 +267,30 @@ You can modify the `problem_type` via problem_type in the Hugging Face config. T
...
+logits *= logit_scale   # scale logits
+logits = activation(logits)  # e.g. sigmoid
+```
+
+Note: `logit_bias` is **subtracted** from the logits (not added), consistent with the `sigmoid_normalize` convention where `sigmoid(x - bias)` centers the sigmoid around the bias value.
+
+| Parameter | Default | Description |
+| --------- | ------- | ----------- |
+| `logit_bias` | `None` | Bias subtracted from logits before activation |
+| `logit_scale` | `None` | Scale factor applied to logits after bias subtraction |
```
Comment (noooop) [inline docs/models/pooling_models/classify.md:298]: Affine Score Calibration, also known as Platt Scaling (Platt, 1999), is the most widely used method for calibrating classifier outputs into well-calibrated probabilities.

The calibration follows the transformation:

`activation(logit_scale * (logit - logit_bias))`

| Parameter | Default | Description |
| --------- | ------- | ----------- |
| `logit_bias` | `None` | Bias subtracted from logits before activation |
| `logit_scale` | `None` | Scale factor applied to logits after bias subtraction |

Note: `logit_bias` is **subtracted** from the logits (not added), consistent with the `sigmoid_normalize` convention where `sigmoid(x - bias)` centers the sigmoid around the bias value.

The computation order is as follows:

```python
logits -= logit_bias    # subtract bias (center scores)
logits *= logit_scale   # scale logits
logits = activation(logits)  # e.g. sigmoid
```

Example configuration:

```bash
--pooler-config '{"use_activation": true, "logit_bias": 4.5, "logit_scale": 1.0}'
```

Comment (noooop) [reply docs/models/pooling_models/classify.md:298]: How about we deprecate `logit_bias` and use `logit_mean` and `logit_sigma` (It would be better to find better names.) to align with LayerNorm?

https://docs.pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html

image

`activation((logit - logit_mean)/logit_sigma)`


```python
logits -= logit_mean  # subtract center scores
logits /= logit_sigma  # divided by the standard-deviation
logits = activation(logits)  # e.g. sigmoid
```

Comment (noooop) [reply docs/models/pooling_models/classify.md:298]: If the name is not changed, most people will think the formula is  `activation(logit_scale * (logit + logit_bias))`

File: docs/models/pooling_models/classify.md (hunk 3 above)
Comment (jefp) [reply docs/models/pooling_models/classify.md]: Done. Renamed to "Affine Score Calibration".

File: docs/models/pooling_models/classify.md (hunk 4 above)
Comment (jefp) [reply docs/models/pooling_models/classify.md:298]: Done. Adopted your structure — Platt scaling intro first, table, note on subtraction, computation order, then example.

Comment (jefp) [reply docs/models/pooling_models/classify.md:298]: This makes sense as a direction but would expand this PR's scope significantly — it deprecates `logit_bias` which is already in production with mxbai-rerank and JinaVL configs. Happy to do the rename + deprecation in a follow-up PR. Would you prefer it in this PR or separate? Also, which target version for the deprecation notice and removal — v0.22?

Comment (jefp) [reply docs/models/pooling_models/classify.md:298]: Agreed on the naming concern. The docs now explicitly state subtraction with a note. The rename to `logit_mean`/`logit_sigma` in a follow-up PR would resolve the ambiguity permanently.

Comment (noooop) [reply docs/models/pooling_models/classify.md:298]: > How about we deprecate `logit_bias` and use `logit_mean` and `logit_sigma` (It would be better to find better names.) to align with LayerNorm?
> 
> https://docs.pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html
> 
> image
> `activation((logit - logit_mean)/logit_sigma)`
> 
> ```python
> logits -= logit_mean  # subtract center scores
> logits /= logit_sigma  # divided by the standard-deviation
> logits = activation(logits)  # e.g. sigmoid
> ```

cc @DarkLight1337 

Do we need to deprecate logit_bias and use logit_mean and logit_sigma in this PR?

Comment (DarkLight1337) [reply docs/models/pooling_models/classify.md:298]: Let's handle the deprecation separately then

**Review by noooop — APPROVED** [review]
thanks!

File: docs/models/pooling_models/classify.md (hunk 4 above)
Comment (jefp) [reply docs/models/pooling_models/classify.md:298]: Once this is merge I will create another PR with the deprecation notice @DarkLight1337
